libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt

```diff
@@ -0,0 +1,392 @@
+        }
+        return eventJson
+    }
+
+    private fun uploadHistory() {
+        if (currentHistoryRecordingSessionState is Idle) {
+            return
+        }
+        endedAt = currentUtcTime()
+        val diffTime = System.currentTimeMillis() - startSessionTime
+        push(DriveEndsEvent(DriveEnds(DriveEndsType.CanceledManually.type, diffTime)))
+        val drive = buildNavigationSession()
+        stopRecording { historyFilePath ->
+            if (hasFeedback || !shouldSendHistoryOnlyWithFeedback) {
+                pushHistoryFile(historyFilePath, drive)
+            }
+        }
+    }
+
+    private fun buildNavigationSession(): CopilotMetadata {
+        val context = mapboxNavigation.navigationOptions.applicationContext
+        return CopilotMetadata(
+            if (isAppDebuggable()) {
+                "mbx-debug"
+            } else {
+                "mbx-prod"
+            },
+            driveMode,
+            driveId,
+            startedAt,
+            endedAt,
+            retrieveNavSdkVersion(),
+            retrieveNavNativeSdkVersion(),
+            context.packageManager.getPackageInfo(
+                mapboxNavigation.navigationOptions.applicationContext.packageName,
+                0,
+            ).versionName,
+            appUserId,
+            appSessionId,
+        )
+    }
+
+    private fun isAppDebuggable(): Boolean {
+        val context = mapboxNavigation.navigationOptions.applicationContext
+        return context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE != 0
+    }
+
+    private fun pushHistoryFile(historyFilePath: String, drive: CopilotMetadata) {
+        val metadata = buildAttachmentMetadata(drive)
+        val uploadOptions = buildUploadOptions(historyFilePath, metadata)
+        HistoryUploadWorker.uploadHistory(
+            mapboxNavigation.navigationOptions.applicationContext,
+            drive,
+            uploadOptions,
+            metadata.sessionId,
+        )
+    }
+
+    private fun buildAttachmentMetadata(copilotMetadata: CopilotMetadata): AttachmentMetadata {
+        val filename = generateFilename(copilotMetadata)
+        val owner = retrieveOwnerFrom(accessToken)
+        val sessionId = generateSessionId(copilotMetadata, owner)
+        return AttachmentMetadata(
+            name = filename,
+            fileId = "",
+            format = GZ,
+            type = ZIP,
+            sessionId = sessionId,
+        )
+    }
+
+    private fun buildUploadOptions(
+        filePath: String,
+        metadata: AttachmentMetadata,
+    ): UploadOptions {
+        val metadataList = arrayListOf(metadata)
+
+        // we have to leave it up to end users to ensure the filename in the metadata matches the actual name of the file
+        val from = File(filePath)
+        val to = copyToAndRemove(from, metadata.name)
+        var url = if (isAppDebuggable()) {
+            STAGING_BASE_URL
```

Comment:
Is it correct for external apps to send data to a staging host?

Reply:
This is a really great question 🤔 

Could we restrict to Nav SDK debuggable?

Reply:
Maybe we can rely on `BuildConfig.DEBUG` from the SDK (`libnavigation-core`) that only applies to us (as published artifacts are `release`), and then, the rest will be sent to the `production` endpoint. What do you think?

Reply:
It think it is fine to use `BuildConfig.DEBUG`. But we should probably discuss this with the team that owns these hosts. Also I don't know how useful traces from debuggable apps usually are.

Reply:
This is pretty useful for testing purposes. IMO we should keep it.

Cc @DorkMatter